### PR TITLE
Add snapshot conversion logic for new binary odsp snapshot format

### PR DESCRIPTION
**Pull request by jatgarg**
Refer: https://github.com/microsoft/FluidFramework/issues/4589

Here is the implementation for conversion logic of new odsp snapshot format:
1.) It converts compact binary representation to IOdspSnapshot(trees, blobs and ops).
2.) It also converts IOdspSnapshot to compact binary representation.
Currently this is standalone logic  and not used to actually convert data from network call of trees latest.
Also the compact format is not yet finalized. So did this with this format: https://github.com/microsoft/FluidFramework/issues/4589#issuecomment-856440945
 
Also includes unit tests for the same. 
Thanks to @vladsud for the reference PR.

### CODE REVIEW

**Comment by jatgarg** [comment]
@sunnyhuang2008 You can also take a look.

**Comment by vladsud** [comment]
Did not yet look at implementation, but a comment based on description:
Using IOdspSnapshot as an entry point into the system is tempting (as it means no changes to existing code), but long term that would add one more format and conversion we do not need.
Instead I'd suggest to create a layer that talks in language that runtime understands (ISnapshotsTree & blobs cache) and make old and new code produce that format.
Some conversion to IOdspSnapshot might be useful to ensure we can roundrip data back and forward without loosing anything.

File: packages/drivers/odsp-driver/src/snapshotFormatConverter.ts [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+const minReadVersion = "1.0";
+
+/**
+ *  Header of the downloaded snapshot.
+ */
+export interface ISnapshotHeader {
+    // This is the minimum version of the reader required to read the wire format of a given snapshot.
+    minReadVersion: string,
+    // Represents the version with which snapshot is created.
+    createVersion: string,
```
Comment (vladsud) [inline packages/drivers/odsp-driver/src/snapshotFormatConverter.ts:20]: That's probably more a question to @sunnyhuang2008 & William:
I'm not sure I follow precise meaning of these two version fields (the reference to reader in particular)
I believe what we are trying to do here is mention format version used, using semantic versioning equivalent - major & minor numbers.

File: packages/drivers/odsp-driver/src/snapshotFormatConverter.ts [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+ */
+export class SnapshotFormatConverter {
+    constructor() {}
+
+    /**
+     * Recreates header section and then validate that.
+     * @param node - tree node to read header section from
+     */
+    private readAndValidateHeaderSection(node: NodeCore) {
+        assert(node.getString(0) === "MinReadVersion", "MinReadVersion header should be present");
```
Comment (vladsud) [inline packages/drivers/odsp-driver/src/snapshotFormatConverter.ts:37]: We should probably allow properties to be written in any order. I.e. have generic code that can read a property bag, and then this code that checks that certain properties are available.

File: packages/drivers/odsp-driver/src/snapshotFormatConverter.ts [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+        assert(node.getString(2) === "CreateVersion", "CreateVersion header should be present");
+        assert(node.getString(4) === "SnapshotSequenceNumber", "SnapshotSequenceNumber header should be present");
+        const header = {
+            minReadVersion: node.getString(1),
+            createVersion: node.getString(3),
+            snapshotSeqNumber: node.getNumber(5),
+        };
+        assert(minReadVersion >= header.minReadVersion, "Driver min read version should >= to server minReadVersion");
```
Comment (vladsud) [inline packages/drivers/odsp-driver/src/snapshotFormatConverter.ts:45]: We should consider version range that client understands being communicated to server as part of request, such that server could support different versions. This does not change the need on client to validate that client can understand output

File: packages/drivers/odsp-driver/src/snapshotFormatConverter.ts [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+    /**
+     * Recreates blobs section of the tree.
+     * @param node - tree node to read blob section from
+     * @param dictionary - name map, used to decode path/IDs.
+     */
+    private readBlobSection(node: NodeCore, dictionary: string[]) {
+        const blobs: Map<string, Uint8Array> = new Map();
+        for (const [idIndex, blob] of node.iteratePairs()) {
+            assert(typeof idIndex === "number", "Blob index should be a number");
+            assert(blob instanceof BlobCore, "Blob content should be of type blob");
```
Comment (vladsud) [inline packages/drivers/odsp-driver/src/snapshotFormatConverter.ts:72]: We should think how to make these asserts actionable. I.e. if it is hit in production the way it's written, it's not really actionable. I.e. we need to convert them to exceptions with enough metadata that gets logged into telemetry, such that we can make some progress using it (or can keep adding metadata until it's clear where the problem is)

File: packages/drivers/odsp-driver/src/snapshotFormatConverter.ts [hunk 5]
```diff
@@ -0,0 +1,279 @@
...
+    }
+
+    /**
+     * Recreates snapshot tree out of tree representation.
+     * @param treeNode - tree node to de-serialize from
+     * @param dictionary - name map, used to decode path/IDs.
+     */
+    private readTreeSection(treeNode: NodeCore, dictionary: string[]) {
+        const tree: ISnapshotTree = {
+            id: "id",
```
Comment (vladsud) [inline packages/drivers/odsp-driver/src/snapshotFormatConverter.ts:99]: id is optional now, let's not add random ids.

File: packages/drivers/odsp-driver/src/snapshotFormatConverter.ts [hunk 6]
```diff
@@ -0,0 +1,279 @@
...
+            assert(typeof pathIndex == "number", "Tree index should be a number");
+            const path = dictionary[pathIndex];
+            assert(path !== undefined, "Path should not be undefined");
+
+            if (child instanceof NodeCore) {
+                tree.trees[path] = this.readTreeSection(child, dictionary);
+            } else {
+                assert(typeof child == "number", "Should be number to look in dictionary");
+                const id = dictionary[child];
+                assert(id !== undefined, "Blob id should not be undefined");
```
Comment (vladsud) [inline packages/drivers/odsp-driver/src/snapshotFormatConverter.ts:114]: I think the right message here - id is out of range, as we get undefined if we read outside of range of array in JS

File: packages/drivers/odsp-driver/src/snapshotFormatConverter.ts [hunk 7]
```diff
@@ -0,0 +1,279 @@
...
+
+    /**
+     * Converts ODSP snapshot format to binary compact representation.
+     * @param snapshotTree - snapshot tree to serialize
+     * @param blobs - blobs to serialize
+     * @param snapshotSeqNumber - seq number at which snapshot is created.
+     * @param ops - ops to serialize
+     * @returns - ReadBuffer - binary representation of the data.
+     */
+    public convertOdspSnapshotToCompactSnapshot(
```
Comment (vladsud) [inline packages/drivers/odsp-driver/src/snapshotFormatConverter.ts:238]: Given this class has no state, consider removing class and exporting only public functions.
I'd probably separate functionality into multiple files - one file for one conversion direction and one - for different direction (maybe 3rd - shared code, if any). With assumption that one direction will not be used in production, and thus can be eliminated by tree shaking. As right now, tree shaking process would leave this function (as it's public method of a class that is used) even if it's not used in production code.

File: packages/drivers/odsp-driver/src/snapshotRepresentation.ts [hunk 8]
```diff
@@ -0,0 +1,486 @@
+/*!
```
Comment (vladsud) [inline packages/drivers/odsp-driver/src/snapshotRepresentation.ts:1]: Please add reference to a ZipIt (I think that's the name) spec and any other interesting resources that might be useful.
We should see if it can be made public, as otherwise it's not very clear what all of this is about. Even if it's not, we need to make it clear here that there is some standard that this code needs to follow and thus structure can't be easily changed.

Also I feel like file name does not reflect what this is about. It has less to do with snapshot and more with specific serialization format and programming model exposing data in such format. So file should be named appropriately

File: packages/drivers/odsp-driver/src/snapshotRepresentation.ts [hunk 9]
```diff
@@ -0,0 +1,486 @@
...
+export abstract class BlobCore {
+    public abstract get buffer(): Uint8Array;
+    /**
+     * Represents a blob.
+     * @param useStringMarkerCode - Represents is the string marker code should be used when representing or
+     * the byte marker code.
+     */
+    constructor(private readonly useStringMarkerCode: boolean = false) {}
+
+    public toString(encoding = stringEncoding) {
```
Comment (vladsud) [inline packages/drivers/odsp-driver/src/snapshotRepresentation.ts:250]: So in my original exploration I chose to support any encoding with utf-8. But I'm not sure what's the story with encodings in ZipIt format. Does it have same lazy approach to encoding? Or it mandates specific encoding? If it's the later, then we should not allow callers to overwrite it.

File: packages/drivers/odsp-driver/src/snapshotRepresentation.ts [hunk 10]
```diff
@@ -0,0 +1,486 @@
...
+    }
+}
+
+/**
+ * "deep copy" blob, holds to binary data passed in
+ * It is called deep copy as a counter-part to BlobShallowCopy, which
+ * is a reference to underlying binary stream (ReadBuffer).
+*/
+class BlobDeepCopy extends BlobCore {
+    constructor(protected readonly data: Uint8Array, useStringMarkerCode: boolean = false) {
```
Comment (vladsud) [inline packages/drivers/odsp-driver/src/snapshotRepresentation.ts:275]: Can you please add comment with @param description of each argument?
It's a bit hard to follow when and why useStringMarkerCode should be true or false. Is this developer choice? Is it mentioned somewhere in payload what need to be used in a given response?

File: packages/drivers/odsp-driver/src/snapshotFormatConverter.ts (hunk 1 above)
Comment (sunnyhuang2008) [reply packages/drivers/odsp-driver/src/snapshotFormatConverter.ts:20]: Correct, the minReadVersion is the format version that the client can safely read without breaking. For example, if we were to add some features to the format, the write version would be lets say v2, but because there is no breaking change, we will specify the min read version to be the same (lets say v1) and the client will be able to still read with a v1 reader without breaking but just won't get the features of v2 without upgrading the reader. It was inspired by the "version made by" and "Version needed to Extract" from the Zip format: https://pkware.cachefly.net/webdocs/casestudies/APPNOTE.TXT

File: packages/drivers/odsp-driver/src/snapshotFormatConverter.ts (hunk 2 above)
Comment (will-tao) [inline packages/drivers/odsp-driver/src/snapshotFormatConverter.ts:37]: >MinReadVersion [](http://example.com/codeflow?start=38&length=14)

test comment

Comment (will-tao) [inline packages/drivers/odsp-driver/src/snapshotFormatConverter.ts:37]: >MinReadVersion [](http://example.com/codeflow?start=38&length=14)

should these be a const variable like minReadVersion above as this is reused here and below in the writeHeaderSection function?

File: packages/drivers/odsp-driver/src/snapshotFormatConverter.ts (hunk 1 above)
Comment (vladsud) [reply packages/drivers/odsp-driver/src/snapshotFormatConverter.ts:20]: It would be great to follow up here and see if <major>.<minor> version solves the problem, as it's standard way to deal with these problems.
Zip file format is rather old, and I think the world moved on to leveraging semantic versioning.

File: packages/drivers/odsp-driver/src/snapshotFormatConverter.ts (hunk 2 above)
Comment (jatgarg) [reply packages/drivers/odsp-driver/src/snapshotFormatConverter.ts:37]: Replaced it with common variable.

Comment (jatgarg) [reply packages/drivers/odsp-driver/src/snapshotFormatConverter.ts:37]: Replaced it with switch.

File: packages/drivers/odsp-driver/src/snapshotFormatConverter.ts (hunk 7 above)
Comment (jatgarg) [reply packages/drivers/odsp-driver/src/snapshotFormatConverter.ts]: Removed the class.

File: packages/drivers/odsp-driver/src/snapshotFormatConverter.ts (hunk 4 above)
Comment (jatgarg) [reply packages/drivers/odsp-driver/src/snapshotFormatConverter.ts]: I think we can defer this work until the actual format is implemented since we don't know these would remain or not.

File: packages/drivers/odsp-driver/src/snapshotFormatConverter.ts (hunk 3 above)
Comment (jatgarg) [reply packages/drivers/odsp-driver/src/snapshotFormatConverter.ts]: Hmm. Will talk to server folks about it.

File: packages/drivers/odsp-driver/src/snapshotRepresentation.ts (hunk 10 above)
Comment (jatgarg) [reply packages/drivers/odsp-driver/src/snapshotRepresentation.ts]: Yes, we have different codes whether the content that follows is utf8 string format or just the general byte[] representation. Added comments and also changed the name. This does not matter when we read the snapshot. It is just used while writing which tells us whether we want to use encode in utf8 string or not.

File: packages/drivers/odsp-driver/src/snapshotRepresentation.ts (hunk 9 above)
Comment (jatgarg) [reply packages/drivers/odsp-driver/src/snapshotRepresentation.ts]: Removed as we only bothers about converting to utf8 string here. zipIt can encode utf8 or base64 but we only want utf8 here.

File: packages/drivers/odsp-driver/src/snapshotRepresentation.ts (hunk 8 above)
Comment (jatgarg) [reply packages/drivers/odsp-driver/src/snapshotRepresentation.ts]: Yes, I will check it out whether we can make it public or not. It is a sharepoint doc.

File: packages/drivers/odsp-driver/src/compactSnapshotParser.ts [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+}
+
+/**
+ * Recreates header section and then validate that.
+ * @param node - tree node to read header section from
+ */
+function readAndValidateHeaderSection(node: NodeCore) {
+    let minReadVersion: string | undefined;
+    let createVersion: string | undefined;
+    for (const [headerName, value] of node.iteratePairs()) {
```
Comment (vladsud) [inline packages/drivers/odsp-driver/src/compactSnapshotParser.ts:32]: Consider if we can make it more generic / reusable, to be useful in other cases where we might have property bags.
For example, if we assume that every property would be of string type, then I'd rather have generic function that takes a node and returns JS object with property bag. Maybe have another function that takes an object and validates that it has certain required properties (passed as an array).

File: packages/drivers/odsp-driver/src/compactSnapshotParser.ts [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+function readBlobSection(node: NodeCore, dictionary: string[]) {
+    const blobs: Map<string, ArrayBuffer> = new Map();
+    for (const [idIndex, blob] of node.iteratePairs()) {
+        assert(typeof idIndex === "number", "Blob index should be a number");
+        assert(blob instanceof BlobCore, "Blob content should be of type blob");
+        const blobId = dictionary[idIndex];
+        assert(blobId !== undefined, "blob id should be present");
+        const unit8Array = blob.buffer;
+        blobs.set(blobId,
+            unit8Array.buffer.slice(unit8Array.byteOffset, unit8Array.byteOffset + unit8Array.byteLength));
```
Comment (vladsud) [inline packages/drivers/odsp-driver/src/compactSnapshotParser.ts:85]: Isn't it equivalent to just using unit8Array? I thought Uint8Array conforms to ArrayBuffer signature, doesn't it?
also unit -> uint?

File: packages/drivers/odsp-driver/src/compactSnapshotParser.ts [hunk 13]
```diff
@@ -0,0 +1,156 @@
...
+
+    readAndValidateHeaderSection(root.getNode(0));
+
+    const dictionary = readDictionarySection(root.getNode(2));
+
+    const tree = readTreeSection(root.getNode(1), dictionary);
+
+    const blobs = readBlobSection(root.getNode(3), dictionary);
+
+    if (root.length === 5) {
```
Comment (vladsud) [inline packages/drivers/odsp-driver/src/compactSnapshotParser.ts:151]: I'd rather make ops sub-tree required, but empty if we do not have ops.
That will allow additions to this tree in the future (if we want to have something else here, i.e. 6th element)

File: packages/drivers/odsp-driver/src/compactSnapshotParser.ts [hunk 14]
```diff
@@ -0,0 +1,156 @@
...
+ * @param buffer - Compact snapshot to be parsed into tree/blobs/ops.
+ * @returns - tree, blobs and ops from the snapshot.
+ */
+export function parseCompactSnapshotResponse(buffer: ReadBuffer) {
+    const builder = TreeBuilder.load(buffer);
+    let ops: ISequencedDeltaOpMessage[] | undefined;
+
+    assert(builder.length === 1, "1 root should be there");
+    const root = builder.getNode(0);
+    assert(root.length >= 4 && root.length <= 5, "4 or 5 sections should be there");
```
Comment (vladsud) [inline packages/drivers/odsp-driver/src/compactSnapshotParser.ts:141]: I'd not limit it from top (5), to allow extensibility in the future. As long as bottom 5 are there, and signatures match, we should consume it.
Alternative design is possible and should be considered - every sub-tree has name (i.e. "trees", "blobs", "ops"), such that order does not matter, and then additions are simple in the future. I think we should strongly consider it.

File: packages/drivers/odsp-driver/src/compactSnapshotWriter.ts [hunk 15]
```diff
@@ -0,0 +1,140 @@
...
+import { ISequencedDeltaOpMessage } from "./contracts";
+import { NodeCore, ReadBuffer, TreeBuilder } from "./zipItDataRepresentationUtils";
+
+/**
+ * Writes header section of the snapshot.
+ * @param node - tree node to serialize to
+ * @param snapshotSeqNumber - seq number at which snapshot is created.
+*/
+function writeHeaderSection(node: NodeCore, snapshotSeqNumber: number) {
+    node.addString("MinReadVersion");
```
Comment (vladsud) [inline packages/drivers/odsp-driver/src/compactSnapshotWriter.ts:18]: For readability, consider creating function addProperty(node, propertyName, propertyValue) and calling it here.

File: packages/drivers/odsp-driver/src/compactSnapshotWriter.ts [hunk 16]
```diff
@@ -0,0 +1,140 @@
...
+        mappingNode.addString(id);
+        i++;
+    }
+
+    writeTree(treeSectionNode, snapshotTree, mapping);
+    // Next is blobs node.
+    const blobsNode = rootNode.addNode();
+    writeBlobsSection(blobsNode, blobs, mapping);
+    // Then write the ops node.
+    if (ops) {
```
Comment (vladsud) [inline packages/drivers/odsp-driver/src/compactSnapshotWriter.ts:135]: I'd make it unconditional

File: packages/drivers/odsp-driver/src/zipItDataRepresentationUtils.ts [hunk 17]
```diff
@@ -0,0 +1,496 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { assert, IsoBuffer, Uint8ArrayToString } from "@fluidframework/common-utils";
+
+/** Default encoding for strings */
+const stringEncoding = "utf-8";
```
Comment (vladsud) [inline packages/drivers/odsp-driver/src/zipItDataRepresentationUtils.ts:9]: given format is only using utf-8, I'd inline this variable - results in slightly smaller bundle size (only 2 usages).

File: packages/drivers/odsp-driver/src/zipItDataRepresentationUtils.ts [hunk 18]
```diff
@@ -0,0 +1,496 @@
...
+        assert(length >= 0, "Skip length should be positive");
+        this.index += length;
+    }
+}
+
+/**
+ * Buffer class, used to sequentially writ data.
+ * Used by tree code to serialize tree into binary representation.
+ */
+class WriteBuffer {
```
Comment (vladsud) [inline packages/drivers/odsp-driver/src/zipItDataRepresentationUtils.ts:65]: Please take a look if there is reasonable way to partition code and separate writing code into separate .ts file.
That way tree shaking can eliminate it as we will not use it (only in tests), but tree shaking is dump and pulls whole file.
It might be tricky as it's not just one class, but methods like BlobCore.write(), TreeBuilder.serialize().
If possible, we should partition these classes into read-only versions, and extensions (possibly stand-alone functions) that provide writing capabilities. It might be challenging, as it might require more things to be made public :(

Comment (vladsud) [reply packages/drivers/odsp-driver/src/zipItDataRepresentationUtils.ts:65]: That said, it's possible that it's Ok to leave it as is with assumption that at some point in future we would want to move summarization process to use similar more compact format, and thus writing capabilities would be required in shipping. So we should do splitting if it makes sense / feels natural, otherwise get a bit more data.

File: packages/drivers/odsp-driver/src/compactSnapshotParser.ts (hunk 12 above)
Comment (jatgarg) [reply packages/drivers/odsp-driver/src/compactSnapshotParser.ts:85]: No. unitArray.buffer is specifically of type ArrayBuffer. We can put uint8 in var of type ArrayBuffer but it would fail the "instanceof" check.

File: packages/drivers/odsp-driver/src/snapshotFormatConverter.ts (hunk 1 above)
Comment (sunnyhuang2008) [reply packages/drivers/odsp-driver/src/snapshotFormatConverter.ts]: I see, with semantic versioning we can resolve the problem as well, with major versions as breaking changes, minor and patch versions as non breaking. The client would have to just be on the same major version to be able to read properly. In this case, the property bag will have string "version" that contains the semantic version of the format instead of the two minRead and CreateVersions. + @William Tao and @Marcelo Mas to comment if there are any other concerns with semantic versions.

File: packages/drivers/odsp-driver/src/snapshotFormatConverter.ts (hunk 3 above)
Comment (sunnyhuang2008) [reply packages/drivers/odsp-driver/src/snapshotFormatConverter.ts]: If we are going with semantic versioning, an idea might be give us the semantic major version that you would like to read and we will return the newest version that's within that major version? like: give us 2 and we will return the newest version that's at least 2.0.0 but no more than 3.0.0, this way it doesn't break the client reading and it uses only one version instead of a version range. If you need more specific version, just specify the minor version, if all three versions are specified, we will give you that exact version.

File: packages/drivers/odsp-driver/src/compactSnapshotParser.ts (hunk 14 above)
Comment (jatgarg) [reply packages/drivers/odsp-driver/src/compactSnapshotParser.ts:141]: But it should be change in the server response that we get? Like currently we just rely on ordering in a  particular way and we have no way of knowing which section is which.

File: packages/drivers/odsp-driver/src/zipItDataRepresentationUtils.ts (hunk 18 above)
Comment (jatgarg) [reply packages/drivers/odsp-driver/src/zipItDataRepresentationUtils.ts:65]: I did put the read and write in different files just for separation of concerns. Otherwise we do inter dependent in a lot of things. Also those seems to be a small code, so it should not affect that much.

File: packages/drivers/odsp-driver/src/compactSnapshotParser.ts (hunk 13 above)
Comment (jatgarg) [reply packages/drivers/odsp-driver/src/compactSnapshotParser.ts:151]: Is it a suggestion for server response? Because we just parse what we get, Currently ops are not necessary.

File: packages/drivers/odsp-driver/src/compactSnapshotWriter.ts (hunk 15 above)
Comment (jatgarg) [reply packages/drivers/odsp-driver/src/compactSnapshotWriter.ts:18]: But why always have extra code which will go through some kind of switch statement whenever we need to add a property. Also we would be adding extra code. Imo this seems readable enough?

File: packages/drivers/odsp-driver/src/compactSnapshotParser.ts (hunk 14 above)
Comment (vladsud) [reply packages/drivers/odsp-driver/src/compactSnapshotParser.ts:142]: Yep. We should not pass on opportunities here - if we agree it's good change, then we should make it on both client & server.

File: packages/drivers/odsp-driver/src/compactSnapshotWriter.ts (hunk 15 above)
Comment (vladsud) [reply packages/drivers/odsp-driver/src/compactSnapshotWriter.ts:19]:
```typescript
const addProperty(node, a, b) => { node.addString(a); node.addString(b); }

function writeHeaderSection(node: NodeCore, snapshotSeqNumber: number) {
   addProperty(node, "MinReadVersion", snapshotMinReadVersion);
   addProperty(node, "CreateVersion", snapshotMinReadVersion);
   addProperty(node, "SnapshotSequenceNumber", snapshotSeqNumber);
}
```

Happy to bet that it will be not bigger in minimized version :)
That's because addProperty would become a single letter, but public names (like addString) stay as is.
So 4 "addString" are removed, bug small function is added.

File: packages/drivers/odsp-driver/src/compactSnapshotParser.ts (hunk 11 above)
Comment (jatgarg) [reply packages/drivers/odsp-driver/src/compactSnapshotParser.ts]: Added an api that will parse and check that a node has certain required properties. I think the other suggestion about adding  some generic function that returns JS object of generic props may not be of that much importance as then we would be iterating over props twice as we don't know what they would be. However it makes sense for known props and I added an api for that.

File: packages/drivers/odsp-driver/src/snapshotFormatConverter.ts (hunk 3 above)
Comment (jatgarg) [reply packages/drivers/odsp-driver/src/snapshotFormatConverter.ts]: This sounds great @sunnyhuang2008 . I think we can proceed with this.

File: packages/drivers/odsp-driver/src/compactSnapshotParser.ts (hunk 14 above)
Comment (jatgarg) [reply packages/drivers/odsp-driver/src/compactSnapshotParser.ts:124]: I have talked to them about adding "string" properties in front of sections like blobs, trees, ops etc. After that we don't need to rely on length.

File: packages/drivers/odsp-driver/src/compactSnapshotWriter.ts (hunk 16 above)
Comment (jatgarg) [reply packages/drivers/odsp-driver/src/compactSnapshotWriter.ts:133]: Currently lets leave it as is because it needs to be in sync with what actual response we are reading from server. Once they make it non optional, then we can reflect that here. Also, the write is currently not used anywhere.

**Comment by jatgarg** [comment]
Ping

File: packages/drivers/odsp-driver/src/compactSnapshotParser.ts [hunk 19]
```diff
@@ -0,0 +1,139 @@
...
+    SnapshotSequenceNumber: number,
+}
+
+/**
+ * Recreates header section and then validate that.
+ * @param node - tree node to read header section from
+ */
+function readAndValidateHeaderSection(node: NodeCore) {
+    const header =
+        getAndValidateNodeProps<ISnapshotHeader>(node, ["MinReadVersion", "CreateVersion", "SnapshotSequenceNumber"]);
```
Comment (vladsud) [inline packages/drivers/odsp-driver/src/compactSnapshotParser.ts:32]: I do not think this cast is buying us anything - this function should just return Record<> and let callers do appropriate casting.
That said, can we use some TS trickery to get a list of property names out of type (ISnapshotHeader) and not have developer generate them manually as an array? Can you please run it by Tony?

File: packages/drivers/odsp-driver/src/compactSnapshotParser.ts (hunk 12 above)
Comment (vladsud) [reply packages/drivers/odsp-driver/src/compactSnapshotParser.ts:73]: I'd suggest moving this logic directly into BlobCore, or into stand-alone static function.
I do not see any way to do shallow copy here :(
Which suggests that our usage of ArrayBuffer for driver API may not have been good choice - Uint8Array would have been a better choice. I guess something to think about for future improvements :)

File: packages/drivers/odsp-driver/src/zipItDataRepresentationUtils.ts [hunk 20]
```diff
@@ -0,0 +1,416 @@
...
+     * Represents a blob.
+     * @param useUtf8Code - Represents if the utf8 string marker code should be used when representing.
+     */
+    constructor(private readonly useUtf8Code: boolean = false) {}
+
+    public toString() {
+        return Uint8ArrayToString(this.buffer, "utf-8");
+    }
+
+    public write(buffer: WriteBuffer) {
```
Comment (vladsud) [inline packages/drivers/odsp-driver/src/zipItDataRepresentationUtils.ts:170]: I think I asked that question earlier: is there way to refactor code such that all write/serialize code is in separate file, for tree shaking to remove it?

File: packages/drivers/odsp-driver/src/zipItDataRepresentationUtils.ts [hunk 21]
```diff
@@ -0,0 +1,416 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
```
Comment (vladsud) [inline packages/drivers/odsp-driver/src/zipItDataRepresentationUtils.ts:4]: Can you please add some reference to ZipIt documentation here?

**Review by vladsud — APPROVED** [review]

File: packages/drivers/odsp-driver/src/zipItDataRepresentationUtils.ts (hunk 21 above)
Comment (jatgarg) [reply packages/drivers/odsp-driver/src/zipItDataRepresentationUtils.ts:4]: Done

File: packages/drivers/odsp-driver/src/zipItDataRepresentationReadUtils.ts [hunk 22]
```diff
@@ -0,0 +1,56 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
```
[the remark this answers is not in the document]
Comment (jatgarg) [reply packages/drivers/odsp-driver/src/zipItDataRepresentationReadUtils.ts:2]: Renamed

File: packages/drivers/odsp-driver/src/zipItDataRepresentationUtils.ts [hunk 23]
```diff
@@ -0,0 +1,416 @@
...
+ * Binary blob is one of three types supported as a leaf node of a tree.
+ * Note: concrete implementations (derived classes) are not exposed from this module
+ */
+export abstract class BlobCore {
+    public abstract get buffer(): Uint8Array;
+    /**
+     * Represents a blob.
+     * @param useUtf8Code - Represents if the utf8 string marker code should be used when representing.
+     */
+    constructor(private readonly useUtf8Code: boolean = false) {}
```
[the remark this answers is not in the document]
Comment (jatgarg) [reply packages/drivers/odsp-driver/src/zipItDataRepresentationUtils.ts:169]: So we do supply this value in:
    public addBlob(blob: Uint8Array, useUtf8Code: boolean = false) {
        this.children.push(new BlobDeepCopy(blob, useUtf8Code));
    }
So we can have 2 types of blobs where we want to represent string or buffer data. Like ops are strings. So we need to supply the zip it code we want to use to represent for the data. So when we write the blob for ops, we use the string code.

File: packages/drivers/odsp-driver/src/compactSnapshotParser.ts (hunk 19 above)
Comment (jatgarg) [reply packages/drivers/odsp-driver/src/compactSnapshotParser.ts]: Removed the cast and getting the record from the utility function. We can not achieve the functionality of getting names as list from the interface.

File: packages/drivers/odsp-driver/src/zipItDataRepresentationUtils.ts (hunk 20 above)
Comment (jatgarg) [reply packages/drivers/odsp-driver/src/zipItDataRepresentationUtils.ts]: Refactored the code to separate the concerns.
